### ISSUE TITLE
debug without (re-)compile

### ISSUE THREAD

**Issue by GitMensch** [OlegKunitsyn/gnucobol-debug#63]
I've now tested this extension with some bigger programs where the source is on a slow network share.
The result: the compilation part takes quite some time, but that is "ok" because out of the scope of this extension.

The part that is really bad: when using "restart" the complete source is re-compiled for no good reason (I didn't changed anything, just wanted to start debugging from scratch).
Please fix that.

In general it would be nice to only compile upon explicit request - maybe have mutliple targets "recompile and debug" + "debug" - or even better: only do the compilation part if the expected binary is either missing or has a time-stamp older than the sources (as this is "unknown" for copybooks it may be reasonable to be able to adjust this behavior in an extension setting or to still provide two targets).

**Comment by brunopacheco1** [comment]
> The part that is really bad: when using "restart" the complete source is re-compiled for no good reason

If you change the field `cobargs` to something like `["-x", "-C"]`, does it recompile the executable or just compiles to C? Or would you say even this compilation to C is an issue?

**Comment by GitMensch** [comment]
> Or would you say even this compilation to C is an issue?

Yes, I'd say. The compile to C looks for all sources specified, then reads all copybooks and does a complete parsing and code generation.
I guess the extension will need the COBOL source files and gdb will need the generated C files so both should be present and up-to-date. Maybe the easiest way is to drop the hard cobc invocation and instead going for make?

**Comment by brunopacheco1** [comment]
I understand the idea of not compiling the binary and C files every time, but to be clear on this issue, the extension needs all of them, the binary, the C files and the COBOL file.

The first to tell GDB what to run, and both C and COBOL files, to match variables names and lines in C to the correspondent in COBOL. When attaching, The executable is not relevant, but the rest is still required by the extension.

The rebuild on restart is definitely something to fix, but anything else the extension has to make sure the expected files are there (namely the COBOL files). That is why the field `group` is relevant, with the COBOL file names the extension is able to look for the C files and with everything in place it does properly the matching step.

We could add a validation step to look for the C files and the binary (when it is not an attaching request). We could add a flag to the configuration to skip the cobc execution for advanced users and explain how a `preLaunchTask` can be set up (I didn't find yet a way to enforce a `preLaunchTask` directly from the extension). @GitMensch what do you think?

**Comment by OlegKunitsyn** [comment]
“There are roads which must not be followed, armies which must not be attacked, towns which must not be besieged, positions which must not be contested, commands of the sovereign which must not be obeyed.” ... and feature requests which must not be implemented. :)

**Comment by GitMensch** [comment]
Concerning the COBOL and C sources: of course they must exist for this extension to work as expected but it's possible in any way it should be done when necessary, not up front.

GDB already knows which files it debugs and we should be able to get the information from it, sometimes the files are not found, in this case GSB will omit a warning and we don't need to care. Otherwise whenever a new C source is recognized it would be scanned "is this a COBOL module?" and if yes all COBOL sources mentioned there would be scanned, if there is an error we can raise the same warning as GDB and just go on  This way it only matters what the binary is and possibly where the C files have to be searched for (actually we only need to setup a path when they were moved and pass that to GDB on start, the actual path would be resolved by GDB) and can search for the COBOL dukes in the current vscode workspace. I'd rather set this up before debugging than to additional setup those paths in the debug extension or even worse specify all the 2000+ programs that may be part of the run in a "group" setting.

**Comment by brunopacheco1** [comment]
If I understood your idea correctly, you are saying we should scan the COBOL and C source files to build the `Source Map` (variables and line references) on-demand, in a sense that, if there is a new file in current GDB scope, we look for it and then let the extension do its job?

**Comment by brunopacheco1** [comment]
If this is the idea, MI has two commands, one to get the current file and another to list all available source files for the executable. `-file-list-exec-source-file` and `-file-list-exec-source-files` respectively.

To keep to current architecture, the second command would be more suitable as it list all files at once, we would just have to filter the listed files to the ones present in the current CWD.

To use the first command, we still can look for the target files at the debug session launching but load the rest of the source files on-demand.

The first approach is heavier but we throw exceptions before starting the debug session. The second approach postpones loading errors and adds overhead to the stack frame evaluation (when there are more files to load).

On both solutions, we would avoid having a list of COBOL files in the launch configuration, though. So I would go for one of them anyway.

My only concern now is, what should we do when the source file is not found? As we would not have a list of files to look for, should we ignore the ones that are not present in the current CWD? For instance, if the user compiles the COBOL to the binary but it doesn't have generated the C source files.

**Comment by GitMensch** [comment]
In general: yes that was the idea. This should solve some issues we see for debugging complete COBOL applications and for attaching to a running process.

I don't know how to apply this best but I'd recommend to not do something different for the first program, and then it is also coming to not start with a COBOL main program but with a module runner similar to "cobcrun" (so you don't have COBOL sources for the target either).

I'm the case of C (it COBOL) sources not found: just do what GDB normally does: show the source name and position in a warning (debug console, also when bit verbose) and let the user either go on or post load the sources via gdb commands in the debug console.
And then it is useful to be able to pre set the src path related variables in GDB via it's variables now tunable as this extension's setting.

I'm totally aware that placing GDB in control of the source paths make it more complicated but as this part is in control of the binary in the first place and used for the stack it would be consequent and in the end more useful to use it for the source file references wherever possible.

Thank you for working on this, I'm quite positive that this is the main missing bit to be able to use the extension for production code.

**Comment by GitMensch** [comment]
@brunopacheco1 Just a friendly ping: any update yet and/or the option to work on this soon?

**Comment by GitMensch** [comment]
As I just did another test of this on Win32 (now working) - the auto-recompile on each debug is really bad. I highly suggest (again) that this extension should _never_ build the programs on their own but use a pre-build-task. This automatically provides (without this extension doing anything of this on its own):

* all compiler outputs shown in the terminal view (without any "verbose" setting)
* depending on the setup of the task can easily be parsed and issues shown in the problem pane
* if debugging cannot start because of compile issues the task fails and vscode shows a clear error message (without any "verbose" setting)
* the task definition _can_ use `make` which would be useful for big projects but also only recompile when necessary

**Comment by GitMensch** [comment]
@ddeclerck @nberth could you tackle this? The combination of compiling and debugging (which kind of breaks "normal" programmer flow) is the one big issue this extension and therefore your fork (as well as @marcsosduma's cobgdb) has.

The development environments I work with are about 3-6 thousand programs (all shared objects), those are all compiled with `-g`; I can always attach to them via GDB - but to do the same with these extensions I need to place all of those in a compilation group and have them recompiled on every debugging session.
That takes hours (and maybe some programs are broken because of a current copybook change, which likely provides a problem for this setup) - and with Windows that even won't work if I'd compile to the same place (as you cannot overwrite DLLs in use there).
